### DISCUSSION TITLE
Example ESPNOW_Sensor will not compile

### DISCUSSION THREAD

**Discussion by hindle49** [timmbogner/Farm-Data-Relay-System#222] [Q&A] [answered]
Hi,

I’ve been following the FDRS project for quite some time and thought it was time I tried it for myself. However, I’ve been having quite some difficulty in compiling the different parts. Quite frustrating on my part, as I can see everything should work from other people’s experiences, so the problem is entirely at my end, but I can’t see how.

Having spent time looking through my setup and reading the on-line discussions I'm left asking for help.

So far I have managed to configure and compile the 0_MQTT_Gateway example. This is working, connects to my Wifi and I can see an initial 'Publication' in my MQTT Broker.

The 1_UART has been giving my some challenges, but today I thought I’d get the “ESPNOW_Sonsor built (direct from the Example) and at least I’d have that sending regular data towards a UART_GW.

Unfortunately, after many attempts, I still can’t get the “ESPNOW_Sensor” to compile without error.

All the libraries in the Arduino IDE 2.3.2 are up to date.

My hardware is a AZ-Delivery ESP32-WROOM-32 module.

I have had the board type set to “ESP32 Dev Module” and “WEMOS D1 MINI ESP32”, and others for experimental purposes. 

I’ve not changed the example code in any way. The only changes I’ve made have been to the fdrs_globals.h file.

Any help would be most gratefully received as I’m lost as to way I’m getting the follow error messages.

```
In file included from D:\user2\ian\Documents\Projects\2024\2024-04-26 Environmental Case Mk3\gitfiles\Farm-Data-Relay-System\examples\ESPNOW_Sensor\ESPNOW_Sensor.ino:10:
d:\user2\ian\Documents\Arduino\libraries\Farm_Data_Relay_System_Main\src/fdrs_node.h: In function 'void beginFDRS()':
d:\user2\ian\Documents\Arduino\libraries\Farm_Data_Relay_System_Main\src/fdrs_node.h:125:28: error: invalid conversion from 'void (*)(const uint8_t*, const uint8_t*, int)' {aka 'void (*)(const unsigned char*, const unsigned char*, int)'} to 'esp_now_recv_cb_t' {aka 'void (*)(const esp_now_recv_info*, const unsigned char*, int)'} [-fpermissive]
  125 |   esp_now_register_recv_cb(OnDataRecv);
      |                            ^~~~~~~~~~
      |                            |
      |                            void (*)(const uint8_t*, const uint8_t*, int) {aka void (*)(const unsigned char*, const unsigned char*, int)}
In file included from d:\user2\ian\Documents\Arduino\libraries\Farm_Data_Relay_System_Main\src/fdrs_node_espnow.h:5,
                 from d:\user2\ian\Documents\Arduino\libraries\Farm_Data_Relay_System_Main\src/fdrs_node.h:67:
C:\Users\ian\AppData\Local\Arduino15\packages\esp32\tools\esp32-arduino-libs\idf-release_v5.1-442a798083\esp32/include/esp_wifi/include/esp_now.h:157:54: note:   initializing argument 1 of 'esp_err_t esp_now_register_recv_cb(esp_now_recv_cb_t)'
  157 | esp_err_t esp_now_register_recv_cb(esp_now_recv_cb_t cb);
      |                                    ~~~~~~~~~~~~~~~~~~^~
Multiple libraries were found for "WiFi.h"
  Used: C:\Users\ian\AppData\Local\Arduino15\packages\esp32\hardware\esp32\3.0.1\libraries\WiFi
  Not used: D:\user2\ian\Documents\Arduino\libraries\WiFiNINA
exit status 1

Compilation error: exit status 1
```

**Comment by timmbogner** [comment]
Hi @hindle49,
This issue is caused by a change made to the ESP32 Arduino core, starting with v3.0.0. It's already been patched on FDRS, so make sure you have the latest version of the repo installed in your Arduino 'libraries' folder. Let me know if that fixes things!

**Reply by timmbogner** [reply, marked as answer]
I think this brings up something else that's been on the to-do list for a while:
FDRS isn't in the Arduino Library Registry. There isn't a good reason, I just thought I'd have to start using a complicated versioning system. Now I see that it's actually pretty easy, and it would have allowed me to automatically patch this issue as it arose without users being inconvenienced. I'll try to get that done soon, I just need to change some wording and then submit it to the official repo.

**Reply by hindle49** [reply]
Tim,
Many thanks for your help. I've not had time to follow up on your feedback, but now that I know where I'm going wrong I'm sure I'll get it to compile when I get chance.

Once again, this is a great project and I'm sure from all the feedback, many people are enjoying its capabilities.

Thanks again
Ian
